'download' should accept full Exercism URLs

verg opened the pull request:
Enhancement from: https://github.com/exercism/cli/issues/341

Allows `exercism download http://exercism.io/submissions/45aaa580ea91452bb34076ade70511f2` and `exercism download exercism.io/submissions/45aaa580ea91452bb34076ade70511f2`

kytrinyx commented on cmd/download.go:
```diff
@@ -13,6 +13,11 @@ import (
 
+const (
+	submission_url_prefix        = "exercism.io/submissions/"
+	submission_url_prefix_w_http = "http://" + submission_url_prefix
```
Would it be possible to use a precompiled regex here that optionally includes http:// (and, potentially, https://)?

kytrinyx commented on cmd/download.go:
```diff
@@ -27,7 +32,8 @@ func Download(ctx *cli.Context) error {
 	}
 
-	submission, err := client.Download(args[0])
+	submissionID := parseSubmissionID(args[0])
```
We've got a bit of a mess in how we name things, but I think that we've been calling this the submission key rather than the submission ID.

kytrinyx commented on cmd/download.go:
```diff
@@ -66,3 +72,10 @@ func writeFile(path, contents string) error {
+
+// Parses Submission ID from url strings
+// e.g "http://exercism.io/submissions/<SUBMISSION_ID>" --> <SUBMISSION_ID>
+func parseSubmissionID(submission string) string {
```
How about `submissionKeyFromURL` as a name for this? parseSubmissionID sounds like it's parsing the ID, which it's not. A name starting in _parse_ tends to focus on what the function is doing rather than what you get out of it, which can be harder to understand.

The function doesn't need to be commented, since it's not exported.

verg commented:
Hey Katrina,

Thanks for the awesome feedback!

How's this look? Okay to slurp everything at the beginning with the ".*"?


Some testing:
```go
gore> re := regexp.MustCompile(".*(https*://)?exercism.io/submissions/")
&regexp.Regexp{regexpRO:regexp.regexpRO{expr:".*(https*://)?exercism.io/submissions/", prog:(*syntax.Prog)(0xc4200664b0), onepass:(*regexp.onePassProg)(nil), prefix:"", prefixBytes:[]uint8(nil), prefixComplete:false, prefixRune:0, prefixEnd:0x0, cond:0x0, numSubexp:1, subexpNames:[]string{"", ""}, longest:false}, mu:sync.Mutex{state:0, sema:0x0}, machine:[]*regexp.machine(nil)}
gore> re.ReplaceAllString("http://exercism.io/submissions/fbfb626b2b534e18bfc7d04e1363fce8", "")
"fbfb626b2b534e18bfc7d04e1363fce8"
gore> re.ReplaceAllString("https://exercism.io/submissions/fbfb626b2b534e18bfc7d04e1363fce8", "")
"fbfb626b2b534e18bfc7d04e1363fce8"
gore> re.ReplaceAllString("exercism.io/submissions/fbfb626b2b534e18bfc7d04e1363fce8", "")
"fbfb626b2b534e18bfc7d04e1363fce8"
gore> re.ReplaceAllString("wasdlkexercism.io/submissions/fbfb626b2b534e18bfc7d04e1363fce8", "")
"fbfb626b2b534e18bfc7d04e1363fce8"
gore> re.ReplaceAllString("fbfb626b2b534e18bfc7d04e1363fce8", "")
"fbfb626b2b534e18bfc7d04e1363fce8"
```

NobbZ commented:
> `".*(https*://)?exercism.io/submissions/"`

The complexity of the resulting automaton is nearly the same, but isn't `".*(https?://)?exercism.io/submissions/"` more correct? Also since we throw the away the match anyway, `".*(?https*://)?exercism.io/submissions/"` could reduce overhead, since there is no additional matching/backtracking group to remember. Dunno if go-re does know about that particular part of syntax.

verg commented:
@NobbZ Good call. I updated `(https*://)?` to `(https?://)?`.

I'm not very familiar, but Go doesn't appear to support `'(?'`. I get this panic:
```go
panic: regexp: Compile(`.*(?https?://)?exercism.io/submissions/`): error parsing regexp: invalid or unsupported Perl syntax: `(?h`
```
Updated tests: https://play.golang.org/p/oQhgB79AOD

kotp commented:
Will this support both forms?  The "last part of the URL" and the full URL?

I just tried to download a submission with the (reported) cli being up to date, but it is failing.

kytrinyx commented:
It sounds like it would be worth having a test for the regex, since it's been a bit tricky to get it working.

nywilken commented:
> Also since we throw the away the match anyway, ".*(?https?://)?exercism.io/submissions/" could reduce overhead, since there is no additional matching/backtracking group to remember. Dunno if go-re does know about that particular part of syntax.

@NobbZ I believe you are advising that the a non-capturing group be used for matching the http string.  If so @verg the syntax you are looking for is `".*(?:https?://)?exercism.io/submissions/"`. Go uses the Re2 syntax which differs from other RegExp engines. More information on Re2 can be found on [Google's Re2 Wiki](https://github.com/google/re2/wiki/Syntax)

NobbZ commented:
Basically there where 2 proposals in that overall comment:

* Change quantifier `*` after the `s` of `https` to `?` because the former would allow urls like `httpssssss://…`.
* Change the match-group to a non-match group to reduce complexity of the resulting DFA. This is a premature optimisation though. It may be, that it doesn't actually change anything in the internal representation of the compiled regular expression. This is subject to benchmarks. Those benchmarks should measure compile time of both expressions as well as execution times against a couple of strings.

kytrinyx commented:
It's been over a year since we heard from the original author on this one. I'm going to go ahead and close the PR.
